File: overlay/clojure/hydra-kernel/src/main/clojure/hydra/overlay/clojure/lib/equality.clj
(ns hydra.overlay.clojure.lib.equality
  (:require [hydra.overlay.clojure.lib.ordering :refer [generic-compare]]))

;; equal :: a -> a -> Bool
;; Delegates to generic-compare (hydra.overlay.clojure.lib.ordering) -- the
;; established pattern already used by maps.clj/sets.clj/lists.clj/eithers.clj
;; -- rather than a separate implementation, per
;; docs/specification/ordering-and-equality.md: equal a b holds exactly when
;; compare a b is equalTo.
;;
;; This also fixes a bug where a bare top-level (:literal (:decimal ...)) term
;; got the correct scale-sensitive comparison (via a dedicated decimal-term?
;; check, using BigDecimal .equals instead of Clojure's scale-blind numeric-
;; tower `=`), but a decimal NESTED inside a record/map (e.g. a Field's term,
;; or a map value) fell straight through to native `=`, which recurses
;; structurally but compares any BigDecimal leaf it reaches with scale-blind
;; equality ((= 1.10M 1.1M) => true) -- silently wrong for exactly the values
;; docs/specification/ordering-and-equality.md's decimal scale-distinctness
;; provision exists to distinguish. generic-compare's decimal branch is
;; unconditional (used at every recursion depth, not just the top level), so
;; delegating to it is correct at any nesting depth.
(def hydra_overlay_clojure_lib_equality_equal
  "Check if two values are equal."
  (fn [a] (fn [b] (zero? (generic-compare a b)))))

File: overlay/scala/hydra-kernel/src/main/scala/hydra/overlay/scala/lib/equality.scala
package hydra.overlay.scala.lib

object equality:
  // Delegates to ordering.compareTerms (hydra.overlay.scala.lib.ordering) rather than a
  // separate implementation. The previous version special-cased a BARE top-level BigDecimal
  // (scala.math.BigDecimal's own equals/hashCode are scale-BLIND by design, delegating to
  // compareTo, unlike java.math.BigDecimal) but fell to native `==` for everything else --
  // correct for a decimal at the term's own top level, but WRONG for one nested inside a
  // Map/Set/Record/etc.: native `==` recurses structurally into case classes/collections,
  // but that recursion eventually reaches a nested BigDecimal's own scale-blind equals,
  // silently losing scale-distinctness (docs/specification/ordering-and-equality.md: 1.1 !=
  // 1.10) below the top level (#742), and analogously for floats: NaN != NaN and -0.0 == 0.0
  // under native ==, wrong per the same spec's extended totalOrder (#745). compareTerms's
  // BigDecimal and Float/Double branches fire at every recursion depth already (it recurses
  // via itself, not a shallow pre-check), so delegating to it fixes both at any nesting
  // depth, not just the top level.
  def equal[A](x: A)(y: A): Boolean = ordering.compareTerms(x, y) == 0
